provider/mem/mem.go

```diff
@@ -148,25 +170,38 @@ func (a *Alerts) Get(fp model.Fingerprint) (*types.Alert, error) {
 
 // Put adds the given alert to the set.
 func (a *Alerts) Put(alerts ...*types.Alert) error {
+	errs := &types.MultiError{}
 
 	for _, alert := range alerts {
 		fp := alert.Fingerprint()
 
+		existing := false
+
 		// Check that there's an alert existing within the store before
 		// trying to merge.
 		if old, err := a.alerts.Get(fp); err == nil {
+			existing = true
+
 			// Merge alerts if there is an overlap in activity range.
 			if (alert.EndsAt.After(old.StartsAt) && alert.EndsAt.Before(old.EndsAt)) ||
 				(alert.StartsAt.After(old.StartsAt) && alert.StartsAt.Before(old.EndsAt)) {
 				alert = old.Merge(alert)
 			}
 		}
 
+		if err := a.callback.PreStore(alert, existing); err != nil {
+			errs.Add(err)
+			continue
+		}
+
 		if err := a.alerts.Set(alert); err != nil {
+			errs.Add(err)
```

Comment:
Thanks for checking `notifier.Manager`. I am wondering if it would make sense to return 400 error in this case... reason being that client is unlikely to succeed on retries, unless some alerts expire in the meantime. 500 may indicate to client that it's worth retrying.

Reply:
I think returning 4xx/5xx based on the type of error (whether it's retryable or not) and having the client correctly reacting on it may make more sense, but I'm wondering if we should keep it as a separate change and have this PR never returning error from `Alerts.Put()` (just log errors) as it's in `master`.

To my understanding, you're going to use the `PreStore()` to check if the limit has been reached and, if so, an error will be returned. With the current state of the logic, this will trigger a retry as well, which doesn't look ideal to me (but not even fundamentally correct, worst case it will hit the limit again).

Reply:
> I think returning 4xx/5xx based on the type of error (whether it's retryable or not) and having the client correctly reacting on it may make more sense, but I'm wondering if we should keep it as a separate change and have this PR never returning error from Alerts.Put() (just log errors) as it's in master.

This is the safe option, I'm fine changing the code this way.